Reclassify some accounts as school-managed for CAP

carl-codeorg opened the pull request:
Google and Microsoft-based accounts are now considered school-managed if they meet one of the following criteria:

- The account was created via a roster sync operation
- The account is in a section

This PR accommodates these exceptions by changing the `Policies::ChildAccount.personal_account?` method. Previously, the method recognized any auth option other than Clever and LTI as personal. Now, the method considers Google and Microsoft accounts to be conditionally school owned, if the user is either in a section or roster synced.

## Links

[JIRA for roster-synced students](https://codedotorg.atlassian.net/browse/P20-1244)
[JIRA for sectioned students](https://codedotorg.atlassian.net/browse/P20-1230)

## Testing story

Refactored the existing spec-style test to a matrix-style test using the same pattern as the other tests in the file that have a long list of test cases. I realize this is a bit backwards, but the test matrix is significantly more concise. Open to suggestions and can look into refactoring it if anyone feels strongly.

<!-- Other aspects to consider. Delete any sections that are not relevant to your change. -->

## Deployment strategy

## Follow-up work

<!--
  List (ideally with Jira links) any clean-up or technical debt that will be addressed in future work.
-->

## Privacy

<!--
  1.	Does this change involve the collection, use, or sharing of new Personal Data?
  2.	Does this change involve a new or changed use or sharing of existing Personal Data?
-->

## Security

<!-- Link to Jira task(s) where sensitive security issues are discussed privately. -->

## Caching

## PR Checklist:

<!--
  The final step! Before you create your PR, double-check that everything is in order.
  Change [ ] to [X] during creation to check boxes.
-->

- [ ] Tests provide adequate coverage
- [ ] Privacy and Security impacts have been assessed
- [ ] Code is well-commented
- [ ] New features are translatable or updates will not break translations
- [ ] Relevant documentation has been added or updated
- [ ] User impact is well-understood and desirable
- [ ] Pull Request is labeled appropriately
- [ ] Follow-up work items (including potential tech debt) are tracked and linked

carl-codeorg commented on dashboard/test/lib/policies/child_account_test.rb:
```diff
@@ -36,7 +36,8 @@ class ComplianceStateTest < ActiveSupport::TestCase
       [[:non_compliant_child, :not_U13], true],
       [[:non_compliant_child, :migrated_imported_from_clever], true],
-      [[:non_compliant_child, :with_lti_auth], true],
+      [[:non_compliant_child, :with_lti_auth], false],
```
This expected test result was changed to false because the changes to `personal_account?` revealed that it was actually incorrect. The test user is created with a password by default, which makes this a personal account, and therefore non-compliant because of the `:non-compliant_child` trait. I added another item to the matrix to test what the old one intended to test: a user a with an LTI auth option but no email login or password.

carl-codeorg commented on dashboard/test/lib/policies/child_account_test.rb:
```diff
@@ -81,13 +82,13 @@ class ComplianceStateTest < ActiveSupport::TestCase
+      [[:non_compliant_child, :migrated_imported_from_google_classroom, :without_encrypted_password, {created_at: '2024-06-29T23:59:59MDT'}], false],
+      [[:non_compliant_child, :migrated_imported_from_google_classroom, :without_encrypted_password, {created_at: '2024-07-01T00:00:00MDT'}], false],
+      [[:non_compliant_child, :with_google_authentication_option, :without_encrypted_password, {created_at: '2024-06-29T23:59:59MDT'}], true],
+      [[:non_compliant_child, :with_google_authentication_option, :without_encrypted_password, {created_at: '2024-07-01T00:00:00MDT'}], false],
```
Added the `:without_encrypted_password` trait here too, because these would all be considered personal accounts with a password.

daynew replied:
`:without_email_auth_option` doesn't need to be called? One thing I've been wondering while going through this PR is email auth option vs the encrypted password...

nicklathe reviewed (approved):
LGTM

mgc1194 reviewed (approved):
The changes look good.
The testing matrix it's a bit confusing to read through but It seems the most efficient way to test all these complex scenarios.

carl-codeorg commented:
> The changes look good. The testing matrix it's a bit confusing to read through but It seems the most efficient way to test all these complex scenarios.

Yeah, I also tried writing it out in spec format but it was comically long and unreadable. There might be some way to do it using Rspec shared contexts but in the examples I saw, nothing was as concise as the matrices we already had in the file.

daynew commented on dashboard/lib/policies/child_account.rb:
```diff
@@ -118,23 +118,45 @@ def self.partially_locked_out?(user)
+      return true if user.encrypted_password.present?
+
+      ao_providers = Set.new(user.authentication_options.pluck(:credential_type))
+      return true if ao_providers.empty?
```
This would be unexpected, right? Migrated but they have no Authentication Options?

carl-codeorg replied:
Yes definitely unexpected, but it's an edge case I think could be possible. There are some users out there who are marked as migrated even though the migration didn't fully succeed.

daynew commented on dashboard/lib/policies/child_account.rb:
```diff
@@ -118,23 +118,45 @@ def self.partially_locked_out?(user)
     else
+      return !conditionally_school_managed?(user) if CONDITIONALLY_SCHOOL_OWNED_TYPES.include?(user.provider)
       SCHOOL_OWNED_TYPES.exclude?(user.provider)
     end
```
The duplicated logic in the `else` could be removed if we did a refactor like this above:
```
ao_providers = Set.new([user.provider])
ao_providers = Set.new(user.authentication_options.pluck(:credential_type)) if user.migrated?
```

carl-codeorg replied:
Yep, I tried this and the logic still seems to work. The function looks cleaner without the if statement and duplicate logic for migrated users.

daynew commented on dashboard/test/lib/policies/child_account_test.rb:
```diff
@@ -496,73 +497,37 @@ class ComplianceStateTest < ActiveSupport::TestCase
-      end
+  test '.personal_account?' do
+    # [User traits, Expected result from personal_account?]
+    test_matrix = [
```
Dang this test matrix is getting big. I'm wondering if there is an easier way to organize it so it's harder to miss test cases. 

I think this matrix is missing tests for when a user has two auth options, like Email + Clever, Email + Google + Section. Basically the test that Email always returns `true` despite any other settings.

carl-codeorg replied:
Good point re: `:without_email_auth_option`. I'll add it into the matrix. The way `user_predates_policy?` is currently implemented, it doesn't matter because the function short-circuits before it checks the auth options if the created_at date is prior to the policy coming into effect. But it's still good to include all the correct traits to ensure these tests continue to do what we want them to.

carl-codeorg commented:
Added the missing cases that @daynew noticed to the test matrix. Also reorganized/sorted it by auth option type and added comments to help improve readability.

One factory trait that was causing issues was `:with_lti_auth`. It goes beyond just adding the auth option and also sets properties on the user and destroys all other auth options, making it hard to test in combination with other auth option types. I decided to add a `:with_lti_authentication_option` trait that just adds the auth option, which is consistent in naming and behavior with the other `:with_{some provider}_authentication_option` traits.
